### PR TITLE
GLTF: export all interfaces

### PR DESCRIPTION
**Pull request by Popov72**
See #4103

### CODE REVIEW

**Comment by sebavan** [comment]
Now that they are public, they all need more docs :-)

https://dev.azure.com/babylonjs/ContinousIntegration/_build/results?buildId=6371&view=logs&j=182f86de-37f5-5af4-c789-e92a7c8fa3dc&t=14c2bff8-fe23-5e9c-07aa-9b09ac155f63

**Comment by Popov72** [comment]
Ah, fantastic...

**Comment by bghgary** [comment]
Since it's so easy to export, I'm wondering maybe we don't need to do this until we need them in the exporter. Also, even if we do, the documentation for this seems unnecessary. I'm not sure if it's useful for people to see this in the documentation. Maybe just use `@hidden` for these?

**Comment by sebavan** [comment]
Ok all good for me so let s just close this and the associated issue @bghgary ??? would prevent @Popov72 to write thousands of missing comments and I feel totally aligned on doing it later based on need.

**Comment by bghgary** [comment]
Do we already have any duplicate interface definitions between the loader and exporter? Maybe just fix those if there are any.

**Comment by Popov72** [comment]
I have added the @hidden tags as it was a quick change.

How can I import some types from the loaders into the serializers? Should the webpack config be changed to reference the loader package?

The 3 extensions supported by the serializer are redefining the interfaces found in the loader extensions.

**Comment by sebavan** [comment]
Maybe we should place them in the gltf definition d.ts instead as it is already accessible by both ? as it feels weird to have the extra npm dependency ???

**Comment by Popov72** [comment]
You are speaking of babylonjs-gltf2interface?

Let me know what I should do.

**Comment by sebavan** [comment]
Yup speaking about this one. @bghgary any thoughts ? I would really prefer prevent the dependency for typings only if we can.

**Comment by bghgary** [comment]
I'm okay with putting it in gltf2-interface.

**Comment by sebavan** [comment]
Perfect sounds like a plan to me !!! :-)

**Comment by sebavan** [comment]
LGTM will merge as soon as the tests are ok :-)

**Comment by bghgary** [comment]
Since gltf2-interface is public, it might be a good idea to make the names of the interfaces more consistent. Maybe we should come up with a naming scheme?

File: dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts [hunk 1]
```diff
@@ -905,4 +905,286 @@ declare module BABYLON.GLTF2 {
...
+    }
+
+    /**
+     * Interfaces from the ExtrasAsMetadata extension
+     */
+
+    /** @hidden */
+    interface ObjectWithMetadata {
+        metadata: any;
+    }
```
Comment (bghgary) [inline dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts:949]: This should not be exposed. Not all interfaces are useful externally.

**Comment by bghgary** [comment]
I also would recommend not exposing any interfaces for glTF extensions that are not ratified yet since they are subject to change.

**Comment by sebavan** [comment]
> Since gltf2-interface is public, it might be a good idea to make the names of the interfaces more consistent. Maybe we should come up with a naming scheme?

They look good to me but I am not the best at names, so do not hesitate if you want some renames

> I also would recommend not exposing any interfaces for glTF extensions that are not ratified yet since they are subject to change. Or at least, we should mark them as such.

A comment would be enough in typedoc as exposing them is our path forward between exporters and loaders.

**Comment by Popov72** [comment]
Comments added and metadata interface removed.

I also tried to rationalize the interface names, let me know what do you think.

**Comment by sebavan** [comment]
It looks all perfect to me :-)

**Comment by Popov72** [comment]
The unit tests of the serializers fail with something like:
image
The faulty js code in `postExportNodeAsync` is:
```typescript
const lightType = (
    babylonLight.getTypeID() == Light.LIGHTTYPEID_POINTLIGHT ? IKHRLightsPunctual_LightType.POINT : (
        babylonLight.getTypeID() == Light.LIGHTTYPEID_DIRECTIONALLIGHT ? IKHRLightsPunctual_LightType.DIRECTIONAL : (
            babylonLight.getTypeID() == Light.LIGHTTYPEID_SPOTLIGHT ? IKHRLightsPunctual_LightType.SPOT : null
        )));
```
It seems `IKHRLightsPunctual_LightType` is undefined (it is coming from `import { IKHRLightsPunctual_LightType, IKHRLightsPunctual_LightReference, IKHRLightsPunctual_Light, IKHRLightsPunctual } from "babylonjs-gltf2interface";`)...

When I execute the unit test code in a PG it does work, I don't get this error.

Any idea why it fails? Maybe it's a problem with the setup of the unit tests that does not make "babylonjs-gltf2interface" available to the test code?

**Comment by RaananW** [comment]
> Any idea why it fails? Maybe it's a problem with the setup of the unit tests that does not make "babylonjs-gltf2interface" available to the test code?

This is just a guess (untested) but usually d.ts files are not compiled to javascript. you can use the values during compile time, but not during runtime, as they are not actually there.
Enums with values should be part of a compiled .ts file to be included in the build

File: dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts [hunk 2]
```diff
@@ -905,4 +905,286 @@ declare module BABYLON.GLTF2 {
...
+        bufferView: number;
+        attributes: { [name: string]: number };
+    }
+
+    /**
+     * Interfaces from the KHR_lights_punctual extension
+     */
+
+    /** @hidden */
+    enum IKHRLightsPunctual_LightType {
```
Comment (bghgary) [inline dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts:958]: Try putting `const enum` here.

Comment (Popov72) [reply dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts]: Thanks, trying it!

Comment (sebavan) [reply dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts]: This should fix the tests... I guess :-)

Comment (RaananW) [reply dist/preview release/glTF2Interface/babylon.glTF2Interface.d.ts]: Or that :-)

**Comment by sebavan** [comment]
Looks like it passed, I ll merge as soon as the rest of the pipeline is green and create the nightly :-)

**Pull request merged**